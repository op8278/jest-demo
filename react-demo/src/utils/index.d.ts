import { ShallowWrapper } from 'enzyme'

export function findTestWrapper(wrapper: ShallowWrapper, selector: any): ShallowWrapper
