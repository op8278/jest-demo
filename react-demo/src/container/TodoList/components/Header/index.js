import React, { Component } from 'react'

class Header extends Component {
  constructor(props) {
    super(props)
    this.state = { value: '' }
  }

  handleInputChange = (e) => {
    this.setState({
      value: e.target.value,
    })
  }

  handleInputKeyUp = (e) => {
    const { value } = this.state

    if (e.keyCode === 13 && value) {
      this.props.addUndoItem(value)
      this.setState({
        value: '',
      })
    }
  }

  render() {
    const { value } = this.state
    return (
      <div className="header">
        <div className="header-content">
          TodoList
          <input
            className="header-input"
            placeholder="Todo"
            data-test="input"
            value={value}
            onChange={this.handleInputChange}
            onKeyUp={this.handleInputKeyUp}
          />
        </div>
      </div>
    )
  }
}

export default Header
