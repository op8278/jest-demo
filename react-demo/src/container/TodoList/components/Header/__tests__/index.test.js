import React from 'react'
import { shallow } from 'enzyme'

import Header from '../index'
import { findTestWrapper } from '../../../../../utils'

describe('Header 单元测试', () => {
  describe('初始化', () => {
    test('快照', () => {
      const wrapper = shallow(<Header />)
      expect(wrapper).toMatchSnapshot()
    })

    test('只包含一个 input 框', () => {
      const wrapper = shallow(<Header />)
      const input = findTestWrapper(wrapper, 'input')
      expect(input).toHaveLength(1)
    })

    test('input 框的值默认为空', () => {
      const wrapper = shallow(<Header />)
      const inputElem = findTestWrapper(wrapper, 'input')
      expect(inputElem).toHaveProp('value', '')
    })
  })

  describe('行为', () => {
    test('input 框是一个受控组件', () => {
      const wrapper = shallow(<Header />)
      wrapper.setState({
        value: '123',
      })

      let inputElem = findTestWrapper(wrapper, 'input')
      expect(inputElem).toHaveProp('value', '123')

      wrapper.setState({
        value: '我是input框的值',
      })

      inputElem = findTestWrapper(wrapper, 'input')
      expect(inputElem).toHaveProp('value', '我是input框的值')
    })

    test('input 框的值随用户输入操作改变', () => {
      const wrapper = shallow(<Header />)
      let inputElem = findTestWrapper(wrapper, 'input')
      expect(inputElem).toHaveProp('value', '')

      // 手动触发 onChange 事件,模拟用户输入操作
      inputElem.simulate('change', {
        target: { value: '我是输入的值' },
      })

      inputElem = findTestWrapper(wrapper, 'input')
      expect(wrapper.state()).toMatchObject({
        value: '我是输入的值',
      })
      expect(inputElem).toHaveProp('value', '我是输入的值')
    })

    test('input 框输入回车时,如果没有输入值,则无操作', () => {
      const fn = jest.fn()
      const wrapper = shallow(<Header addUndoItem={fn} />)

      wrapper.setState({ value: '' })
      let inputElem = findTestWrapper(wrapper, 'input')

      inputElem.simulate('keyUp', {
        keyCode: 13,
      })

      expect(fn).toHaveBeenCalledTimes(0)

      inputElem = findTestWrapper(wrapper, 'input')
      expect(wrapper.state()).toMatchObject({
        value: '',
      })
      expect(inputElem).toHaveProp('value', '')
    })

    test('input 框输入回车时,如果有输入值,则 addUndoItem 函数被调用,并且 input 框被清空', () => {
      const fn = jest.fn()

      const wrapper = shallow(<Header addUndoItem={fn} />)
      wrapper.setState({ value: '我是输入内容' })
      let inputElem = findTestWrapper(wrapper, 'input')
      inputElem.simulate('keyUp', {
        keyCode: 13,
      })

      expect(fn).toHaveBeenCalledTimes(1)
      expect(fn).toHaveBeenCalledWith('我是输入内容')

      inputElem = findTestWrapper(wrapper, 'input')
      expect(wrapper.state()).toMatchObject({
        value: '',
      })
      expect(inputElem).toHaveProp('value', '')
    })
  })
})
