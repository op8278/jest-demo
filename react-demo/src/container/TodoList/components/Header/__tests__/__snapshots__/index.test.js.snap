// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header 单元测试 初始化 快照 1`] = `
<div
  className="header"
>
  <div
    className="header-content"
  >
    TodoList
    <input
      className="header-input"
      data-test="input"
      onChange={[Function]}
      onKeyUp={[Function]}
      placeholder="Todo"
      value=""
    />
  </div>
</div>
`;
